File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章防御力.py
#防御力 2018年第九届蓝桥杯国赛  lanqiaoOJ 226



def cmp(x): return x[1]

n1,n2 = map(int,input().split())
a = list(map(int,input().split()))
b = list(map(int,input().split()))
for i in range(n1): a[i]=(i+1,a[i])
for i in range(n2): b[i]=(i+1,b[i])
a.sort(key=cmp)
b.sort(key=cmp,reverse=True)
s = input()
idx1, idx2 = 0, 0
for i in range (n1+n2):
  if s[i]=='1': print("B%d"%b[idx1][0]); idx1 += 1
  else:         print("A%d"%a[idx2][0]); idx2 += 1
print("E")

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章防御力.cpp
//防御力 2018年第九届蓝桥杯国赛  lanqiaoOJ 226 

#include <bits/stdc++.h>
using namespace std;
struct nodea{int id,w;} a[100005];      // A的道具//id是道具，w是这个道具的增加量
struct nodeb{int id,w;} b[100005];      // B的道具
bool cmp1(nodea a,nodea b){
    if(a.w != b.w) return a.w < b.w;    //先对A的增加量排序，从小到大
    else return a.id < b.id;            //再按字典序id排序
}
bool cmp2(nodeb a,nodeb b){             //先对B的增加量排序，从大到小
    if(a.w != b.w) return a.w > b.w;
    else return a.id < b.id;
}
int main(){
    int n1,n2;   cin >> n1 >> n2;
    for(int i = 1; i <= n1; i++)  cin >> a[i].w,  a[i].id = i;
    for(int i = 1; i <= n2; i++)  cin >> b[i].w,  b[i].id = i;
    sort(a+1,a+n1+1,cmp1);
    sort(b+1,b+n2+1,cmp2);
    string s;    cin >> s;
    int idx1 = 1, idx2 = 1;
    for(int i=0;i<s.length();i++){
        if(s[i]=='1'){cout << "B"; cout << b[idx1++].id << "\n";}
        else         {cout << "A"; cout << a[idx2++].id << "\n";}
    }
    cout << "E" << "\n";
    return 0;
}
